File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
const direction = {
    'U': [1, 0],
    'D': [-1, 0],
    'R': [0, 1],
    'L': [0, -1]
};

const makeSet = (a, b, c, d, visited) => {
    visited.add(`${a},${b},${c},${d}`);
    visited.add(`${c},${d},${a},${b}`);
}

const checkSet = (a, b, c, d, visited) => {
    return visited.has(`${a},${b},${c},${d}`);
}

function solution(dirs) {
    let ret = 0;  
    let y=5, x=5;
    
    const visited = new Set();
    
    for (const dir of dirs){
        const [dy, dx] = direction[dir];
        const [ny, nx] = [y + dy, x + dx];
        
        const isIn = ny <= 10 && nx <= 10 && ny >= 0 && nx >= 0;
        if(!isIn) continue;
        
        if(!checkSet(y, x, ny, nx, visited)){
            ret++;
            makeSet(y, x, ny, nx, visited);
        }
        y = ny, x = nx;
    }
    
    return ret;
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(string dirs) {
    int ret = 0;
    int y=5, x=5; // 시작 위치, 평면이동을 통해 0~10 범위로 이동
    int visited[12][12][12][12]; // 현재 위치, 다음 위치를 통해 경로 방문 처리
    map<char, pair<int,int>> mp;
    
    // 방향 벡터 Map 
    mp['U'] = {1, 0};
    mp['R'] = {0, 1};
    mp['D'] = {-1, 0};
    mp['L'] = {0, -1};
    
    for (char dir : dirs){
        int dy, dx;
        tie(dy, dx) = mp[dir];
        
        int ny = y + dy;
        int nx = x + dx;
        
        // 좌표평면 이탈
        if(ny > 10 || nx > 10 || ny < 0 || nx < 0) continue;
        
        if(!visited[y][x][ny][nx]){
            visited[y][x][ny][nx] = 1;
            visited[ny][nx][y][x] = 1; // 양방향 다 체크해줘야함
            ret++;
        }
        y = ny; 
        x = nx;
    }
    
    return ret;
}